src/Features/Core/Portable/CodeRefactorings/SyncNamespace/IChangeNamespaceService.cs

```diff
@@ -9,6 +8,33 @@ namespace Microsoft.CodeAnalysis.ChangeNamespace
 {
     internal interface IChangeNamespaceService : ILanguageService
     {
-        Task<Solution> ChangeNamespaceAsync(Solution solution, ImmutableArray<DocumentId> documentIds, string declaredNamespace, string targetNamespace, CancellationToken cancellationToken);
+        /// <summary>
+        /// Determine whether we can change the namespace for given <paramref name="container"/> in the document.
+        /// Linked documents are not supported, except for a regular document in a multi-targeting project, 
+        /// where the container node must be consistent among all linked documents.
+        /// Here's the additional requirements on <paramref name="container"/> to use this service:
+        /// 
+        /// - If <paramref name="container"/> is a namespace declaration node:
+        ///    1. Doesn't contain or is nested in other namespace declarations
+        ///    2. The name of the namespace is valid (i.e. no errors)
+        ///    3. No partial type declared in the namespace. Otherwise its multiple declaration will
+        ///       end up in different namespace.
+        ///
+        /// - If <paramref name="container"/> is a compilation unit node:
+        ///    1. It must contain no namespace declaration
+        ///    2. No partial type declared in the document. Otherwise its multiple declaration will
+        ///       end up in different namespace.
+        ///   
+        /// Returns <see langword="true"/> only when all the requirements above are met.
+        /// </summary>
+        Task<bool> CanChangeNamespaceAsync(Document document, SyntaxNode container, CancellationToken cancellationToken);
```

Comment:
is this also checking that hte namespace name doesn't match the file name?  or is that done at a different level?

Reply:
Why namespace would match file name?

Reply:
Basically, i'm trying to figure htis out.  Say i have "namespace A.B.C { }" and i'm in a file called `A\B\C\Foo.cs` would this return 'true' (because it does'nt care about if the namespace is already properly named based on the document), or would it return 'false' (because the namespace is already correct based on the doc.

Reply:
I see. No, this service doesn't care, it will return ``true`. That's the logic left behind in the SyncNamespace refactoring. All this requires to change a namespace name is that the declaration  isn't nested.

Reply:
Thanks. That's probably worth calling out in the doc comment IMO.

Reply:
Not sure about this. This service has nothing to do with file name and there's no mention about it in the code. Adding a doc comment here seems odd, especially for people who don't know how this service is evolved from the sync namespace refactoring.

Reply:
FWIW, as someone reading this, it was the first question that came to mind.  i.e. "here's a service that will change a namespace to match something... so... will it do anything if the namespace already matches the document i gave it?"  I don't think it hurts to have remarks to help clarify things where there are multiple reasonable interpretations that can be made". Up to you.